cpp/include/cudf/lists/detail/gather.cuh

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <thrust/transform_scan.h>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+
+/**
+ * @brief The information needed to create an iterator to gather level N+1
+ *
+ * @ref make_gather_data
+ */
+struct gather_data {
+  // The offsets column from our parent list (level N)
+  std::unique_ptr<column> offsets;
+  // For each offset in the above offsets column, the original offset value
+  // prior to being gathered.
+  // Example:
+  // If the offsets[3] == 6  (representing row 3 of the new column)
+  // And the original value it was itself gathered from was 15, then
+  // base_offsets[3] == 15
+  rmm::device_uvector<int32_t> base_offsets;
+  // size of the gather map that will be generated from this data
+  size_type gather_map_size;
+};
+
+/**
+ * @copydoc cudf::make_gather_data(cudf::lists_column_view const& source_column,
+ *                                 MapItType gather_map,
+ *                                 size_type gather_map_size,
+ *                                 cudaStream_t stream,
+ *                                 rmm::mr::device_memory_resource* mr)
+ *
+ * @param prev_base_offsets The buffer backing the base offsets used in the gather map. We can
+ *                          free this buffer before allocating the new one to keep peak memory
+ *                          usage down.
+ */
+template <bool NullifyOutOfBounds, typename MapItType>
+gather_data make_gather_data(cudf::lists_column_view const& source_column,
+                             MapItType gather_map,
+                             size_type gather_map_size,
+                             cudaStream_t stream,
+                             rmm::mr::device_memory_resource* mr,
+                             rmm::device_uvector<int32_t> prev_base_offsets)
```

Comment:
Technically in your usage this is never getting copied because you're doing a `std::move`. Let's make it explicit that this function expects to take ownership by using an r-value ref. 

```suggestion
                             rmm::device_uvector<int32_t>&& prev_base_offsets)
```